src/LLMProviders/chainManager.ts

```diff
@@ -43,41 +45,26 @@ export default class ChainManager {
   public brevilabsClient: BrevilabsClient;
   public static retrievedDocuments: Document[] = [];
 
-  constructor(
-    app: App,
-    getLangChainParams: () => LangChainParams,
-    encryptionService: EncryptionService,
-    settings: CopilotSettings,
-    vectorStoreManager: VectorStoreManager,
-    brevilabsClient: BrevilabsClient
-  ) {
+  constructor(app: App, vectorStoreManager: VectorStoreManager, brevilabsClient: BrevilabsClient) {
     // Instantiate singletons
     this.app = app;
-    this.langChainParams = getLangChainParams();
-    this.settings = settings;
     this.vectorStoreManager = vectorStoreManager;
-    this.memoryManager = MemoryManager.getInstance(this.getLangChainParams(), settings.debug);
-    this.encryptionService = encryptionService;
-    this.chatModelManager = ChatModelManager.getInstance(
-      () => this.getLangChainParams(),
-      encryptionService,
-      this.settings.activeModels
-    );
+    this.memoryManager = MemoryManager.getInstance();
+    this.chatModelManager = ChatModelManager.getInstance();
     this.embeddingsManager = this.vectorStoreManager.getEmbeddingsManager();
-    this.promptManager = PromptManager.getInstance(this.getLangChainParams());
+    this.promptManager = PromptManager.getInstance();
     this.brevilabsClient = brevilabsClient;
-    this.createChainWithNewModel(this.getLangChainParams().modelKey);
-  }
-
-  public getLangChainParams(): LangChainParams {
-    return this.langChainParams;
-  }
-
-  public setLangChainParam<K extends keyof LangChainParams>(
-    key: K,
-    value: LangChainParams[K]
-  ): void {
-    this.langChainParams[key] = value;
+    this.createChainWithNewModel();
+    subscribeToModelKeyChange(() => this.createChainWithNewModel());
```

Comment:
chain needs to be recreated when model key or settings change.

Here it listens to all settings key changes, which is not necessary. But it will cover all edge cases in case someone adds new settings dependency in the future.